### ISSUE TITLE
RFC: ``From`` implementation for ``Dom``

### ISSUE THREAD

**Issue by Kibouo** [maps4print/azul#184]
Disclaimer: I'm happy to make a PR myself. Don't expect the maintainer to do this under current circumstances :)

## Description
The ``.dom()`` function is practically begging to be a trait. The function definition is always the same, apart from a few widgets requiring extra data for rendering. However, these exceptions are in a minority.

## Describe your solution
I propose implementing ``From`` for each widget for the ``Dom`` struct. For the ones with no extra data this would be straightforward.

My solution for types which require extra data, such as ``TableView`` and ``TextInput``, is adding an intermediate type. This type will store (a reference to) the extra data as well as the widget. It will simply deref to the underlying type for easy out-of-order chaining.

First off every "complex" widget would require a function to create the bundle of extra data and widget:
```Rust
// disregard lifetimes, etc. in this example
// Would a ref to the base-type be better?
struct TextInputWithRenderData {
    text_input: TextInput,
    extra_data: &ExtraDataType
}

impl TextInput {
    pub fn add_render_data(self, extra_data: &ExtraDataType) -> TextInputWithRenderData { //... }
}
```
Then we could use it as follows:
```Rust
TextInput::new()
        .bind(info.window, &self.text_input, &self)
        .add_render_data(&self.text_input)
        .into()
        .with_id("text_input_1");
```

The benefit of this would be that functions do not have to take a ``Dom`` anymore. Instead they can take an ``impl Into<Dom>``. This eliminates the need for repetitive ``.dom()`` calls throughout a user's codebase. (The ``.into()`` in the example would also not be needed to be done by the user, it would be handled by the library methods.) 
It is cleaner and simply a more flexible design.

Examples without ``.dom()``:
```Rust
    let label = Label::new("label");
    // ``.with_callback`` takes an ``impl Into<Dom>`` and thus allows for seamless chaining.
    let button = Button::with_label("button")
            .with_callback(On::MouseUp, Callback(some_fn));

    // the wrapper type created by ``add_render_data()`` derefs to ``TextInput`` so chaining in any order is possible.
    TextInput::new()
        .add_render_data(&self.text_input)
        .bind(info.window, &self.text_input, &self)
        .with_id("text_input_1");

    // ``.with_child()`` again takes an ``impl Into<Dom>``
    let dom = Dom::div()
         .with_child(label)
         .with_child(button);
```

## Drawbacks
One could argue that ``.add_render_data()`` just takes the place of the "conventional function" from ``.dom()``. However, requiring extra data is not the norm. There would be less "loose"/"conventional functions" throughout the codebase compared to having a "loose" ``.dom()`` on *every* widget.  

## Is this a breaking change
If we really want we can keep the old ``.dom()`` functions for backwards compatibility.

**Comment by fschutt** [comment]
No, this won't work. The code example you gave won't compile, because you can't define `.with_callback()` on an `Into<Dom>` trait and you need that for method-chaining. `impl Into<Dom>` is for *return parameters*, not for function arguments. How would the function signature for `.with_callback()` look like?

```rust
let button = Button::with_label("button") // <- you HAVE to call .into() here, since .with_callback() is defined on the Dom struct
        .with_callback(On::MouseUp, some_fn);
```

And at that point, you've basically just replaced `.dom()` with `.into()`, which is both a less descriptive function name and can lead to compile errors (when the compiler can't figure out what generic implementation of `.into()` you want to call, then you have to use something like `.into::<Dom<T>>()`, which is just ugly).

So, *replacing* might not be a good idea, but *adding* `Into<Dom>` implementations makes sense. What you could do is to rework `.with_child()` and `.add_child()` to take an `Into<Dom>`, i.e.

```rust
pub fn add_child<T: Into<Dom>>(&mut self, child: T) {
    self.children.push(child.into()) 
}
```

This way you could write this code:

```rust
Dom::div()
    .with_child(Button::with_label("hello")) // no .dom()
    .with_child(Label::new("Test")) // no .dom()
    .with_child(Dom::image("image.jpg"))
```

.. so that you don't have to type `.dom()` everywhere. That would work. However, this will likely increase compilation time a lot (generic code is generally fairly slow to compile). And for widgets that do take an extra argument or have callbacks attached, you could always use the `.dom()` function.

Implementing `Into<Dom>` doesn't require you to remove the `dom()` function - I agree that for "simple" widgets like Label / Button, etc. you could implement `Into<Dom>` additional to using the `.dom()` function, but not for complex ones like SVG.

`.dom()` is just a convention, similar to classes having a `::new()` function. Originally, I used completely pure, free-standing functions like `render_button()`, `render_label(text) -> Dom`, and all classes like "Button" and "Label" are just wrappers around these functions since I liked the capitalization more.

**Comment by Kibouo** [comment]
Sorry, I glanced over the fact that ``.with_callback()`` is currently directly implemented on ``Dom`` (I expected a trait). You actually can implement traits, and thus functions, on "any ``T`` that implements ``Trait``" (although correctly noted not with the ``impl Trait`` syntax). The implementation would be as follows:
```Rust

```
This *does* compile and pass every test (after small unrelated fixes) on master 6763ca7 (23 Jun) with 1.31.1.
So for this to work we'd simply take all methods of ``Dom`` and create traits for them.

As for compile times, I'm not experienced with what does and what doesn't influence it. Would a dedicated ``IntoDom`` trait be better?

P.S.: ``cargo fmt`` results in a lot of files being changed. Is this normal?

**Comment by fschutt** [comment]
I have my doubts, simply because I'm thinking of "how I be able to port this to a C API" - eventually I want to make python / lua, etc. bindings for azul, so I have to use the C ABI and C doesn't know about `impl Trait`. Second, a goal of the `Dom` is to eventually be able to be a constant function, i.e. so you can declare static parts of a `Dom` as `const MY_BTN:  Dom = Button::new("hello").dom();` - this could work if `Vec::new()` will be a `const fn` ([tracking issue](https://github.com/rust-lang/rust/issues/57563)), but I don't know how this'll work with const fn or the C API.

And it's just a simple `.dom()` function, I really don't want to go overboard with traits here. At least for 0.1, maybe just postpone the issue until the first version is released.

if you want to do PRs, please don't use `cargo fmt`, my code style is a protected species ;) - it's just that the default `cargo fmt` config introduces a lot of newlines, which makes things unnecessarily hard to read, I eventually have to make a proper config file for the code style.

**Comment by Kibouo** [comment]
Certainly a valid reason to not go through with this at the moment. I love traits (and writing less code) but exposing complex trait stuff as a C API certainly isn't straightforward.

I'll check out some issues and to-do's in the meantime. Let's get an initial release before discussing new ideas! :smile: 

(This can stay open as a possibility for the future? Unless you decide otherwise ofc.)
